jobrunner/run.py

```diff
@@ -195,28 +155,142 @@ def handle_running_job(job):
+        with set_log_context(job=job):
+            try:
+                handle_job_api(job, api)
+            except Exception:
+                mark_job_as_failed(job, f"Internal error")
+                # Do not clean up, as we may want to debug
+                #
+                # Raising will kill the main loop, by design. The service manager
+                # will restart, and this job will be ignored when it does, as
+                # it has failed. If we have an internal error, a full restart
+                # might recover better.
+                raise
+
+
+def handle_job_api(job, api):
+    """Handle an active job.
+
+    This contains the main state machine logic for a job. For the most part,
+    state transitions follow the same logic, which is abstracted. Some
+    transitions require special logic, mainly the initial and final states, as
+    well as supporting cancellation.
+    """
+    assert job.state in (State.PENDING, State.RUNNING)
+    code = None
+    message = None
+
+    definition = job_to_job_definition(job)
+    old_status = api.get_status(definition)
+
     if job.cancelled:
-        log.info("Cancellation requested, killing job")
-        api.terminate(job_to_job_definition(job))
+        # cancelled is driven by user request, so is handled explicitly first
+        # regardless of executor state.
+        api.terminate(definition)
+        api.cleanup(definition)
+        mark_job_as_failed(job, "Cancelled by user", StatusCode.CANCELLED_BY_USER)
+        return
 
-    try:
-        is_running = sync_job_status(job, api)
+    if old_status.state == ExecutorState.PENDING:
+        # new job
+
+        # check dependencies
+        awaited_states = get_states_of_awaited_jobs(job)
+        if State.FAILED in awaited_states:
+            mark_job_as_failed(
+                job,
+                "Not starting as dependency failed",
+                code=StatusCode.DEPENDENCY_FAILED,
+            )
+            return
+
+        if any(state != State.SUCCEEDED for state in awaited_states):
+            set_message(
+                job, "Waiting on dependencies", code=StatusCode.WAITING_ON_DEPENDENCIES
+            )
+            return
+
+        next_state = ExecutorState.PREPARING
+        new_status = api.prepare(definition)
+
+    elif old_status.state == ExecutorState.PREPARED:
+        next_state = ExecutorState.EXECUTING
+        new_status = api.execute(definition)
+
+    elif old_status.state == ExecutorState.EXECUTED:
+        next_state = ExecutorState.FINALIZING
+        new_status = api.finalize(definition)
+
+    elif old_status.state == ExecutorState.FINALIZED:
+        # final state - we have we have finished!
+        results = api.get_results(definition)
+        # TODO: implement workspace API
+        # delete_obsolete_files(job, results)
+        save_results(job, results)
+        mark_job_as_completed(job)
+        api.cleanup(definition)
```

Comment:
I think the order is wrong here. We shouldn't mark it as completed until we've triggered cleanup, in case there is an error saving the state to the database.

Reply:
but then if there is a db error, the results will be gone, and we can't recover?

Reply:
as in, the next time this job is considered, api.get_results() will presumably fail?

Reply:
You're quite right. My mistake.